### PR TITLE
Dem insertion distribution object

### PR DESCRIPTION
**Pull request by OGaboriault**
# Description of the problem

- This PR is the first of two. The objective is to implement new types of diameter distributions for the DEM. Currently, the code has the uniform and normal distribution. These distribution are using the same snippet of code. The only different is that the uniform distribution uses a standard deviation equal to zero. Since both distribution are using the same snippet of code, the diameter generation was directly being made in the insert function, which doesn't allows for new distribution types to be implemented. 

# Description of the solution

- The "Insertion" base class has a new "Distribution_base" type attribute. This base class is purely virtual and has a function called "particle_size_sampling" which can be override by derived classes such as "NormalDistribution". 
- The "Insertion" class has a new constructor which is being used by its derived classes for the initialization of the distribution_object. 

# How Has This Been Tested?

- Since this is a refactoring and not a new feature, no new test has been implemented. 
- All current tests pass
- Some examples have been used to visualize the normal distribution. 

# Documentation

- N/A

# Future changes

- We could do a separate class for the uniform distribution, but I don't think this should be a priority. The code work the way it is currently. I my mind, this would just add new files for pretty much nothing.
- This refactoring will enable the implementation of new distribution (mainly log-normal and bar chart)

### CODE REVIEW

File: include/dem/insertion.h
Comment (blaisb) [inline include/dem/insertion.h:134]:
```suggestion
  // Number of particles that is going to be inserted at each insertion
  // step. This value can change in the last insertion step to reach the desired
  // number of particles
```

File: include/dem/normal_distribution.h
Comment (blaisb) [inline include/dem/normal_distribution.h:2]: If you want you can also put all the distributions in the same file. No point in having one file per distribution IMO. This is what we did with the physical properties model

File: include/dem/normal_distribution.h
Comment (blaisb) [inline include/dem/normal_distribution.h:40]: Wait but there are only two params?

File: include/dem/normal_distribution.h
Comment (blaisb) [inline include/dem/normal_distribution.h:49]: Please add the @param for both parameters

**Review by blaisb — CHANGES_REQUESTED** [review]
First set of comments will finish tomorrow

File: source/dem/insertion.cc
Comment (lpsaavedra) [inline source/dem/insertion.cc:44]: I don't understand this.... why are we doing the same if the distribution is uniform and normal? Or is it a WIP?

File: source/dem/volume_insertion.cc
Comment (lpsaavedra) [inline source/dem/volume_insertion.cc:42]:
```suggestion
  // Check to see if the remaining un-inserted particles are equal to zero or not
```

File: tests/dem/insertion_non_uniform.cc
Comment (lpsaavedra) [inline tests/dem/insertion_non_uniform.cc:72]: I am not familiar with this tests but why would you have two tests with uniform distribution? Isn't one for non-uniform and one for uniform? At least this is what I get from the name of the tests...

File: include/dem/volume_insertion.h
Comment (lpsaavedra) [inline include/dem/volume_insertion.h:20]: Is this needed? You are not using anything from the normal distribution in this class, are you?

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
I have only some comments/doubts that need to be addressed before approval

File: source/dem/insertion.cc
Comment (OGaboriault) [reply source/dem/insertion.cc:44]: Good question. An uniform distribution is just a normal distribution with a standard deviation equal to zero. This is how the code was working previously. (If you check in the parsing function,dem_parameters.lagrangian_physical_properties.particle_size_std  is set to zero when the distribution type is uniform)

We could add a new class called UniformDistribution with its own constructor, but I think that this would go in a other PR.

File: tests/dem/insertion_non_uniform.cc
Comment (OGaboriault) [reply tests/dem/insertion_non_uniform.cc:72]: Are you talking about the name of the file? 

Previously (less than a month ago), we had three types of INSERTION methods called uniform, non-uniform and plane. The uniform insertion was not useful since you could obtained the same type of insertion location with the non-uniform one. The uniform insertion method got deprecated and the non-uniform insertion method got renamed volume insertion method. In brief, we could technically rename this test `insertion_volume.cc` and deprecated the `insertion_uniform.cc` 

Now, this PR is about size distribution, so the diameter value at which particles are being inserted. Currently we have two types of size distribution, uniform (every particle with the same diameter) and normal (particle diameters follow a normal distribution). 

Like I said in a previous comment, the difference between uniform and normal size distribution is the standard deviation (equal or not equal to zero). If you check the parsing function in `core/parameter_lagrangian.cc`, line 70, you will see that the `enum class size_distribution_type ` was previously never set to uniform or normal, but now it needs to be and this is why you see this modification in this test. 

I hope this helps.  :-)

File: source/dem/insertion.cc
Comment (lpsaavedra) [reply source/dem/insertion.cc:44]: Somehow I think it would be cleaner but I am not sure... maybe if we do not want to create a new class, then we can use the same one but I would add a comment within this if so it would be clear for anybody reading the code

File: tests/dem/insertion_non_uniform.cc
Comment (lpsaavedra) [reply tests/dem/insertion_non_uniform.cc:72]: I see... in that case I do think that it would be better to rename the files? So that the name of the test actually reflects what is being tested...

File: source/dem/insertion.cc
Comment (blaisb) [reply source/dem/insertion.cc:44]: But then shouldnt at least the standard deviation be fixed to zero?
Maybe we should deprecate uniform completely and just ask the user to write normal with no standard deviation. Seems like a very weird option

File: tests/dem/insertion_non_uniform.cc
Comment (blaisb) [reply tests/dem/insertion_non_uniform.cc:72]: I agree with @lpsaavedra

**Review by blaisb — CHANGES_REQUESTED** [review]
Other comments

File: source/dem/insertion.cc
Comment (OGaboriault) [reply source/dem/insertion.cc:44]: When the distribution is set to unifom, the standard deviation is fixed to zero in the parsing function. The NormalDistribution constructor takes two std::unordered_map as inputs to support the uses of many types of particles, that is why it is set to zero there and not here. (Otherwise, I would have to create a std::unordered_map here, which would have been weird)

I could deprecated the uniform size distribution... not sure if it is a good thing, especially when you consider new types of size distribution that are going to be added in the following weeks.

File: tests/dem/insertion_non_uniform.cc
Comment (OGaboriault) [reply tests/dem/insertion_non_uniform.cc:72]: Ok, will renamed the test. What should I do we the insertion_uniform.cc test? deprecate or renamed to insertion_volume_1.cc and insertion_volume_2.cc ?

Comment (blaisb) [reply tests/dem/insertion_non_uniform.cc:72]: Yeah

File: source/dem/insertion.cc
Comment (blaisb) [reply source/dem/insertion.cc:44]: Well I think it's better to hardcode the parameter in the constructor than in the parser of the parameters
In all cases, uniform should always always have a std-dev of 0, no point in passing a st-dev assuming it has been set to zero previously

File: include/dem/insertion.h
Comment (voferreira) [inline include/dem/insertion.h:132]:
```suggestion
  // Number of particles that are going to be inserted at each insertion
```

File: include/dem/normal_distribution.h
Comment (voferreira) [inline include/dem/normal_distribution.h:33]:
```suggestion
   * @param standard_deviation Standard deviation of the normal distribution.
```

File: include/dem/plane_insertion.h
Comment (voferreira) [inline include/dem/plane_insertion.h:36]:
```suggestion
 * Particle insertion using cells cut by a plane. Locally owned cells that are
```

File: include/dem/plane_insertion.h
Comment (voferreira) [inline include/dem/plane_insertion.h:37]:
```suggestion
  * cut by the plane are flagged. From those flagged cells, we insert a particle at
```

File: include/dem/plane_insertion.h
Comment (voferreira) [inline include/dem/plane_insertion.h:38]:
```suggestion
 * their centroids if they are individually empty (contain no particle). This
```

File: source/core/parameters_lagrangian.cc
Comment (voferreira) [inline source/core/parameters_lagrangian.cc:68]: I would use "distribution function" or "distribution function type" for the parameter instead of using size again here. And for the variable, distribution_function or distribution_type would make more sense imo

File: source/dem/insertion.cc
Comment (voferreira) [reply source/dem/insertion.cc:44]: I like more the idea of having a separate class for the uniform distribution without a standard deviation parameter. This way when other distributions would be implemented, it will be clear that another class need to be implemented and nothing would get mixed.

File: source/dem/insertion.cc
Comment (voferreira) [inline source/dem/insertion.cc:93]: I would use abs instead

**Review by voferreira — COMMENTED** [review]
Good job! Just added some comments. I would really like to have distribution functions separated to have a cleaner and more sustainable structure. This would save time on future implementations of other size distribution functions

File: include/dem/normal_distribution.h
Comment (OGaboriault) [reply include/dem/normal_distribution.h:40]: Yep... I realized afterward that the third parameter wasn't useful and forgot to remove it.

Comment (OGaboriault) [reply include/dem/normal_distribution.h:40]: Got it

File: source/dem/insertion.cc
Comment (OGaboriault) [reply source/dem/insertion.cc:44]: I will do separate class, this will solve the problem.

File: tests/dem/insertion_non_uniform.cc
Comment (OGaboriault) [reply tests/dem/insertion_non_uniform.cc:72]: This was not a yes or no question, but I will assume you said yes to the second option. 🫠

File: source/core/parameters_lagrangian.cc
Comment (OGaboriault) [reply source/core/parameters_lagrangian.cc:68]: Are you talking about the string in the prm.get ?

Comment (voferreira) [reply source/core/parameters_lagrangian.cc:68]: Both

Comment (OGaboriault) [reply source/core/parameters_lagrangian.cc:68]: Ok, I'm not sure if I agree. I think it's fine the way it is right not. 

1. Changing the get string would break compatibility with all the prm file. It's not a problem for us, put for all the users out there it's better not to (except when we don't have the choice).  
2. For a random user, when you say "distribution function" it's not clear what the distribution is about. Like you could be talking about a distribution for the rolling resistance coefficient or something. I think keeping the word "size" or "diameter" is better. 

It could do this if you prefer: `const std::string size_distribution_type_str = prm.get("size_distribution_type");`

Comment (blaisb) [reply source/core/parameters_lagrangian.cc:68]: Although we never use _ in parameter names
we use spaces

Comment (OGaboriault) [reply source/core/parameters_lagrangian.cc:68]: My mistake, you are right. It's a typo

const std::string size_distribution_type_str = prm.get("size distribution type")


I have added a "_str", that is what I was talking about.

Comment (voferreira) [reply source/core/parameters_lagrangian.cc:68]: I got it. Nope, using the type in the variable name is not necessary. You are right to keep it as it is

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:25]: This should be identical to the name of the file:

```suggestion
#ifndef distributions_h
#  define distributions_h
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:63]:
```suggestion
   * @param particle_type The type of particle being inserted.
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:52]:
```suggestion
   * @param d_averages Average diameters for each type of particle.
   * @param d_standard_deviations Standard deviation of the diameter for each type of particle.
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:70]:
```suggestion
  // Average diameters for each particle type.
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:72]:
```suggestion
  // Standard deviation of the diameter for each particle type.
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:81]:
```suggestion
   * The constructor store the parameters necessary to define the uniform
   * distribution
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:92]:
```suggestion
   * @param particle_type The type of particle being inserted.
```

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:99]:
```suggestion
  // Diameter values for each particle type.
```

File: include/dem/insertion.h
Comment (lpsaavedra) [inline include/dem/insertion.h:158]:
```suggestion
  // Stores particles diameters
```

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
Some last small comments from my side

File: include/dem/distributions.h
Comment (lpsaavedra) [inline include/dem/distributions.h:103]:
```suggestion
#endif /* distributions_h */
```

**Review by lpsaavedra — APPROVED** [review]

**Review by voferreira — APPROVED** [review]

**Review by blaisb — APPROVED** [review]

**Pull request merged**
